src/queues/EventRedactionQueue.ts

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2019-2021 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import { LogLevel, MatrixClient } from "matrix-bot-sdk"
+import { ERROR_KIND_FATAL } from "../ErrorCache";
+import { logMessage } from "../LogProxy";
+import { RoomUpdateError } from "../models/RoomUpdateError";
+import { redactUserMessagesIn } from "../utils";
+
+export interface QueuedRedaction {
+    /** The room which the redaction will take place in. */
+    readonly roomId: string;
+    /**
+     * Carries out the redaction task and is called by the EventRedactionQueue
+     * when processing this redaction.
+     * @param client A MatrixClient to use to carry out the redaction.
+     */
+    redact(client: MatrixClient): Promise<any>
+    /**
+     * Used to test whether the redaction is the equivalent to another redaction.
+     * @param redaction Another QueuedRedaction to test if this redaction is an equivalent to.
+     */
+    redactionEqual(redaction: QueuedRedaction): boolean
+}
+
+/**
+ * Redacts all of the messages a user has sent to one room.
+ */
+export class RedactUserInRoom implements QueuedRedaction {
+    userId: string;
+    roomId: string;
+
+    constructor(userId: string, roomId: string) {
+        this.userId = userId;
+        this.roomId = roomId;
+    }
+
+    public async redact(client: MatrixClient) {
+        await logMessage(LogLevel.DEBUG, "Mjolnir", `Redacting events from ${this.userId} in room ${this.roomId}.`);
+        await redactUserMessagesIn(client, this.userId, [this.roomId]);
+    }
+
+    public redactionEqual(redaction: QueuedRedaction): boolean {
+        if (redaction instanceof RedactUserInRoom) {
+            return redaction.userId === this.userId && redaction.roomId === this.roomId;
+        } else {
+            return false;
+        }
+    }
+}
+/**
+ * This is a queue for events so that other protections can happen first (e.g. applying room bans to every room).
+ */
+export class EventRedactionQueue {
+    private toRedact: Map<string, QueuedRedaction[]> = new Map<string, QueuedRedaction[]>();
```

Comment:
Could you document that the `string` is a roomId?

Ideally, some day, we'll introduce a newtype for `roomId` but we haven't done so yet.